cpp/src/arrow/engine/substrait/util.cc

```diff
@@ -0,0 +1,139 @@
+/// bindings for consuming a Substrait plan.
+class SubstraitExecutor {
+ public:
+  explicit SubstraitExecutor(
+      std::shared_ptr<Buffer> substrait_buffer,
+      AsyncGenerator<util::optional<compute::ExecBatch>>* generator,
+      std::shared_ptr<compute::ExecPlan> plan, compute::ExecContext exec_context)
+      : substrait_buffer_(std::move(substrait_buffer)),
+        generator_(generator),
+        plan_(std::move(plan)),
+        exec_context_(exec_context) {}
+
+  Result<std::shared_ptr<RecordBatchReader>> Execute() {
+    for (const compute::Declaration& decl : declarations_) {
+      RETURN_NOT_OK(decl.AddToPlan(plan_.get()).status());
+    }
+    RETURN_NOT_OK(plan_->Validate());
+    RETURN_NOT_OK(plan_->StartProducing());
+    // schema of the output can be obtained by the output_schema
+    // of the input to the sink node.
+    auto schema = plan_->sinks()[0]->inputs()[0]->output_schema();
+    std::shared_ptr<RecordBatchReader> sink_reader = compute::MakeGeneratorReader(
+        schema, std::move(*generator_), exec_context_.memory_pool());
+    return sink_reader;
+  }
+
+  Status Close() { return plan_->finished().status(); }
+
+  Status Init() {
+    if (substrait_buffer_ == NULLPTR) {
+      return Status::Invalid("Buffer containing Substrait plan is null.");
+    }
+    std::function<std::shared_ptr<compute::SinkNodeConsumer>()> consumer_factory = [&] {
+      return std::make_shared<SubstraitSinkConsumer>(generator_);
+    };
+    ARROW_ASSIGN_OR_RAISE(declarations_,
+                          engine::DeserializePlan(*substrait_buffer_, consumer_factory));
+    return Status::OK();
+  }
+
+ private:
+  std::shared_ptr<Buffer> substrait_buffer_;
+  AsyncGenerator<util::optional<compute::ExecBatch>>* generator_;
+  std::vector<compute::Declaration> declarations_;
+  std::shared_ptr<compute::ExecPlan> plan_;
+  compute::ExecContext exec_context_;
+};
+
+arrow::PushGenerator<util::optional<compute::ExecBatch>>::Producer
+SubstraitSinkConsumer::MakeProducer(
+    AsyncGenerator<util::optional<compute::ExecBatch>>* out_gen) {
+  arrow::PushGenerator<util::optional<compute::ExecBatch>> push_gen;
+  auto out = push_gen.producer();
+  *out_gen = std::move(push_gen);
+  return out;
+}
+}  // namespace
+
+Result<std::shared_ptr<RecordBatchReader>> ExecuteSerializedPlan(
+    std::shared_ptr<Buffer> substrait_buffer) {
+  arrow::AsyncGenerator<util::optional<compute::ExecBatch>> sink_gen;
+  ARROW_ASSIGN_OR_RAISE(auto plan, compute::ExecPlan::Make());
+  compute::ExecContext exec_context(arrow::default_memory_pool(),
+                                    ::arrow::internal::GetCpuThreadPool());
+  arrow::engine::SubstraitExecutor executor(std::move(substrait_buffer), &sink_gen,
+                                            std::move(plan), exec_context);
+  RETURN_NOT_OK(executor.Init());
+  ARROW_ASSIGN_OR_RAISE(auto sink_reader, executor.Execute());
+  RETURN_NOT_OK(executor.Close());
```

Comment:
This doesn't seem right to me.  We shouldn't be waiting for the execution plan to finish.  If we are going to do that we may as well return `arrow::Table` or `RecordBatchVector`.  However, if we don't wait for it to close I see we will end up with a new dilemma.  Something needs to keep the `ExecPlan` alive.

Instead, can we have `SubstraitExecutor` extend `RecordBatchReader`?  `Execute` can store the `std::shared_ptr<RecordBatchReader>` as a member variable.  You can just proxy the two virtual methods:

```
std::shared_ptr<Schema> SubstraitExecutor::schema() const override {
  return sink_reader->schema();
}
Status SubstraitExecutor::ReadNext(std::shared_ptr<RecordBatch>* batch) override {
  return sink_reader->ReadNext(batch);
}
```

This should allow us to return `SubstraitExecutor` itself instead of returning `sink_reader`.  We should also, arguably, add a context manager for it in python so that `SubstraitExecutor::Close` is called as soon as python is done with it.  Since `RecordBatchReader` does not have a close method I suppose that will be a new wrinkle in how we expose this to python.

We could either return `std::shared_ptr<RecordBatchReader>` and the "close" could be resetting the shared pointer This would still require some cython logic and probably a new python/cython class (CloseableRecordBatchReader?).

Or we could expose SubstraitExecutor in python.

Or we could create the new concept of a CloseableRecordBatchReader in C++ and python.

Or we could add Close to RecordBatchReader (which has a default implementation of doing nothing).

I think my preference would be for the last option there.  CC @lidavidm for second opinion.

Reply:
@westonpace I see your point here. Thanks for noting this one. Exposing more content to Cython and solving this issue is feasible, but I think it would be much better to go with the option you're suggesting. 
Specially, I am not quite sure exposing the SubstraitExecutor to Python is a scalable thing when we want to do more things with this plugin in future. It would require updating things in C++/Cython/Python. But I could be wrong. 

cc @lidavidm

Reply:
(3) seems the best but my main concern is making sure we also update/migrate all the existing uses of the APIs which will be a long project.

Reply:
Shall I proceed with 3 for this PR?

Reply:
CC @pitrou for opinions too but if we proceed with (3) I don't want the rest of the work to be abandoned.

Reply:
I suppose we can have 'new' implementations of RecordBatchReader that do depend on Close() being called also add a destructor-time check.

Reply:
One thing to clear up, don't we have to update the other interfaces extended from `RecordBatchReader`?
If so, should it be another PR?

Reply:
cc @lidavidm @pitrou @westonpace

Reply:
It can be another PR and JIRA, yes. This one is already involved enough.

Reply:
@pitrou @westonpace @lidavidm Jira: https://issues.apache.org/jira/browse/ARROW-16515